lib/circuit/CircuitState/CircuitState.ts

```diff
@@ -0,0 +1,70 @@
+import { Policy } from '../Policy/Policy';
+
+export class CircuitState {
+
+  private state: 'open' | 'close' | 'half-open' = 'open';
+  private removeExecutionsTimers: number[] = [];
+
+  constructor(
+    private readonly timeout: number,
+    private readonly interval: number,
+    private readonly errorsFilter: (error: Error) => boolean,
+    private readonly policy: Policy,
+  ) { }
+
+  public allowExecution(): boolean {
+    return this.state !== 'close';
+  }
+
+  public register(error?: Error): this {
+    const isError = error && this.errorsFilter(error);
+    const type = isError ? 'error' : 'success';
+
+    if (this.state === 'half-open') {
+      isError ? this.close() : this.open();
```

Comment:
Ternary operator is recommended with assignments, in this case it has side-effects.